src/metadata/metadata_superblock.c

```diff
@@ -0,0 +1,431 @@
+			format"\n", __func__, ##__VA_ARGS__)
+#else
+#define OCF_DEBUG_TRACE(cache)
+#define OCF_DEBUG_PARAM(cache, format, ...)
+#endif
+
+
+/**
+ * @brief Super Block - Set Shutdown Status
+ *
+ * @param shutdown_status - status to be assigned to cache.
+ *
+ * @return Operation status (0 success, otherwise error)
+ */
+void ocf_metadata_set_shutdown_status(ocf_cache_t cache,
+		enum ocf_metadata_shutdown_status shutdown_status,
+		ocf_metadata_end_t cmpl, void *priv)
+{
+	struct ocf_metadata_ctrl *ctrl;
+	struct ocf_superblock_config *superblock;
+
+	OCF_DEBUG_TRACE(cache);
+
+	/*
+	 * Get metadata hash service control structure
+	 */
+	ctrl = (struct ocf_metadata_ctrl *) cache->metadata.priv;
+
+	/*
+	 * Get super block
+	 */
+	superblock = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
+
+	/* Set shutdown status */
+	superblock->clean_shutdown = shutdown_status;
+	superblock->magic_number = CACHE_MAGIC_NUMBER;
+
+	/* Flush superblock */
+	ocf_metadata_flush_superblock(cache, cmpl, priv);
+}
+
+static void ocf_metadata_store_segment(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_metadata_context *context = priv;
+	int segment = ocf_pipeline_arg_get_int(arg);
+	struct ocf_metadata_ctrl *ctrl;
+	ocf_cache_t cache = context->cache;
+	int error;
+
+	ctrl = (struct ocf_metadata_ctrl *)cache->metadata.priv;
+
+	context->segment_copy[segment].mem_pool =
+		env_malloc(ctrl->raw_desc[segment].mem_pool_limit, ENV_MEM_NORMAL);
+	if (!context->segment_copy[segment].mem_pool)
+		OCF_PL_FINISH_RET(pipeline, -OCF_ERR_NO_MEM);
+
+	error = env_memcpy(context->segment_copy[segment].mem_pool,
+			ctrl->raw_desc[segment].mem_pool_limit, METADATA_MEM_POOL(ctrl, segment),
+			ctrl->raw_desc[segment].mem_pool_limit);
+	if (error) {
+		env_free(context->segment_copy[segment].mem_pool);
+		context->segment_copy[segment].mem_pool = NULL;
+		OCF_PL_FINISH_RET(pipeline, error);
+	}
+
+	ocf_pipeline_next(pipeline);
+}
+
+static void ocf_metadata_check_crc_sb_config(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_metadata_context *context = priv;
+	struct ocf_metadata_ctrl *ctrl;
+	struct ocf_superblock_config *sb_config;
+	ocf_cache_t cache = context->cache;
+	int segment = metadata_segment_sb_config;
+	uint32_t crc;
+
+	ctrl = (struct ocf_metadata_ctrl *)cache->metadata.priv;
+	sb_config = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
+
+	crc = env_crc32(0, (void *)sb_config,
+			offsetof(struct ocf_superblock_config, checksum));
+
+	if (crc != sb_config->checksum[segment]) {
+		/* Checksum does not match */
+		ocf_cache_log(cache, log_err,
+				"Loading %s ERROR, invalid checksum",
+				ocf_metadata_segment_names[segment]);
+		OCF_PL_FINISH_RET(pipeline, -OCF_ERR_INVAL);
+	}
+
+	ocf_pipeline_next(pipeline);
+}
+
+static void ocf_metadata_check_crc_if_clean(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	ocf_metadata_check_crc_skip(pipeline, priv, arg, true);
+}
```

Comment:
This could be moved to `metadata_segment_ops.c`. This function is the only reason for exporting `ocf_metadata_check_crc_skip()`, otherwise it could be made static.

Reply:
Btw calculating CRC is dependent on superblock segment and this split between superblock/segment ops/other segments seems to lack ability to reflect this dependency gracefully. Maybe we can find a way to make ops superblock agnostic - e.g. inject dependency to superblock by some abstraction or make other sections superblock aware so that the coupling isn't introduced by ops themselves?

Reply:
done